plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts

```diff
@@ -49,23 +47,92 @@ import {
 import { TeamMember as AdoTeamMember } from 'azure-devops-node-api/interfaces/common/VSSInterfaces';
 import { Logger } from 'winston';
 import { PolicyEvaluationRecord } from 'azure-devops-node-api/interfaces/PolicyInterfaces';
-import { WebApi } from 'azure-devops-node-api';
+import {
+  WebApi,
+  getHandlerFromToken,
+  getPersonalAccessTokenHandler,
+} from 'azure-devops-node-api';
 import {
   TeamProjectReference,
   WebApiTeam,
 } from 'azure-devops-node-api/interfaces/CoreInterfaces';
 import { UrlReader } from '@backstage/backend-common';
+import { Config } from '@backstage/config';
+import {
+  AzureDevOpsCredentialsProvider,
+  DefaultAzureDevOpsCredentialsProvider,
+  ScmIntegrations,
+} from '@backstage/integration';
+import {
+  mappedBuildRun,
+  mappedGitTag,
+  mappedPullRequest,
+  mappedRepoBuild,
+} from './mappers';
 
 /** @public */
 export class AzureDevOpsApi {
-  public constructor(
-    private readonly logger: Logger,
-    private readonly webApi: WebApi,
-    private readonly urlReader: UrlReader,
-  ) {}
+  private readonly logger: Logger;
+  private readonly urlReader: UrlReader;
+  private readonly config: Config;
+  private readonly credentialsProvider: AzureDevOpsCredentialsProvider;
+
+  private constructor(
+    logger: Logger,
+    urlReader: UrlReader,
+    config: Config,
+    credentialsProvider: AzureDevOpsCredentialsProvider,
+  ) {
+    this.logger = logger;
+    this.urlReader = urlReader;
+    this.config = config;
+    this.credentialsProvider = credentialsProvider;
+  }
+
+  static fromConfig(
+    config: Config,
+    options: { logger: Logger; urlReader: UrlReader },
+  ) {
+    const scmIntegrations = ScmIntegrations.fromConfig(config);
+    const credentialsProvider =
+      DefaultAzureDevOpsCredentialsProvider.fromIntegrations(scmIntegrations);
+    return new AzureDevOpsApi(
+      options.logger,
+      options.urlReader,
+      config,
+      credentialsProvider,
+    );
+  }
+
+  private async getWebApi(host?: string, org?: string): Promise<WebApi> {
+    // If no host or org is provided we fall back to the values from the `azureDevOps` config section
+    // these may have been setup in the `integrations.azure` config section
+    // which is why use them here and not just falling back on them entirely
+    const validHost = host ?? this.config.getString('azureDevOps.host');
+    const validOrg = org ?? this.config.getString('azureDevOps.organization');
+    const url = `https://${validHost}/${encodeURI(validOrg)}`;
```

Comment:
`encodeURI` is almost never what you want, most commonly it's `encodeURIComponent`. But it shouldn't be harmful in this particular usage I think 🤔

Reply:
https://github.com/backstage/backstage/pull/21795 addresses this 👍